### PR TITLE
Auto delete unused variations when auto generating variations

### LINKED ISSUE #39687: Delete unused variations after removing variable options
**Description**

Once a user deletes a variation option, we should delete the auto generated variations that include that option. We will notify the user about this.

**Design**

AkVNGImLgSqCObTQ3idVn7-fi-1639%3A216854

Screenshot 2023-08-11 at 10 30 06 AM

**Epic**

#36286

**ACs**

- [ ] Show Delete popup when user tries to remove a value from the Variation options block.
- [ ] If user clicks 'Delete' we should remove the matching variations.

### PR DESCRIPTION
**Pull request by louwie17**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR add's a `delete` prop to the `generate` variations API to allow for auto deleting of unused variations.
The PR also add's a new confirmation modal for when a user tries to delete a variation option.

Closes #39687 

https://github.com/woocommerce/woocommerce/assets/2240960/528aad81-7c15-462e-90e2-6d7a98adf51f

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Requires feature flag: `product-variation-management`**

1. Load this branch and enable the new product editor under **WooCommerce > Settings > Advanced > Features**
2. Install the WooCommerce Beta tester plugin and enable the `product-variation-management` feature under **Tools > WCA Test Helper > Features**
3. Go to **Products > Add New** and create a new product, and add multiple variation options under Variations tab.
4. Once you added multiple variation option it should have auto generated some variations
5. Now Edit one of the variation options and remove one of the selected values ( ex: Color: blue, red remove the 'red' value ) and click `update`
6. Notice how the `Variations` list updated and removed some variations that contained the addition value.
7. Now make sure you have one variation option with three selected values and one with a single selected value ( three auto generated variations should be present )
8. Click the 'x' button on the option that has the three options
9. A modal should popup with `Delete <variation option name>` with a Delete and cancel button
10. Click `Delete`
11. The Variations list should be updated now with just a single item, removing the other 2 variations.
12. You can attempt several other scenarios to see if it always deletes/creates the correct number of variations.
13. Now go to **Organization** tab and add a attribute here
14. After adding the attribute, click 'x' to remove it.
15. It should still show the native 'alert' prompt here instead of the Modal

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/data/src/crud/index.ts [hunk 1]
```diff
@@ -54,13 +52,10 @@ export const createCrudDataStore = ( {
...
 
+	const crudReducer = createReducer( reducer );
+
 	registerStore( storeName, {
-		reducer: reducer
-			? ( combineReducers( {
-					crudReducer,
-					reducer,
-			  } ) as Reducer )
-			: ( crudReducer as Reducer< ResourceState > ),
```
Comment (louwie17) [inline packages/js/data/src/crud/index.ts:63]: `combineReducers` actually broke the store, as `combineReducers` changes the structure of the `state` by add prefixes for each store.
Instead I pass the optional `reducer` into `createReducer` and call it there.

File: packages/js/data/src/product-variations/actions.ts [hunk 2]
```diff
@@ -2,30 +2,78 @@
...
+		yield controls.dispatch(
+			'core',
+			'saveEntityRecord',
+			'postType',
+			'product',
+			{
+				id: urlParameters[ 0 ],
+				...productData,
+			}
+		);
```
Comment (louwie17) [inline packages/js/data/src/product-variations/actions.ts:70]: Moving this here, allows us to have a global state for the entire `generateVariations` process, and allows us to give a much more accurate loading state.

File: packages/js/product-editor/src/blocks/variation-items/editor.scss [hunk 3]
```diff
@@ -1,4 +1,7 @@
-.wp-block-woocommerce-product-variations-items-field {
+.wp-block-woocommerce-product-variation-items-field {
+	@media ( min-width: #{ ($break-medium) } ) {
+		min-height: 420px;
```
Comment (louwie17) [inline packages/js/product-editor/src/blocks/variation-items/editor.scss:3]: Moved from the `variations-table` style sheet. Felled more appropriate here, so the table can just be the exact combined height of the rows within.

File: packages/js/product-editor/src/components/variations-table/variations-table.tsx [hunk 4]
```diff
@@ -60,41 +60,56 @@ export function VariationsTable() {
...
-	if ( ! variations || isLoading ) {
+	if ( ! variations && isLoading ) {
 		return (
-			<div className="woocommerce-product-variations is-loading">
+			<div className="woocommerce-product-variations__loading">
 				<Spinner />
+				{ isGeneratingVariations && (
+					<span>
+						{ __( 'Generating variations...', 'woocommerce' ) }
+					</span>
```
Comment (louwie17) [inline packages/js/product-editor/src/components/variations-table/variations-table.tsx:101]: We are actually able to accurately decipher now if the `generateVariations` action is being triggered. Previously we were not able to differentiate if it was `loading` the variations or `generating` them.

File: packages/js/product-editor/src/hooks/use-product-variations-helper.ts [hunk 5]
```diff
@@ -32,36 +28,32 @@ export function useProductVariationsHelper() {
...
-				const hasVariableAttribute = attributes.some(
-					( attr ) => attr.variation
-				);
-				await saveEntityRecord< Promise< Product > >(
-					'postType',
-					'product',
-					{
-						id: productId,
-						type: hasVariableAttribute ? 'variable' : 'simple',
-						attributes,
```
Comment (louwie17) [inline packages/js/product-editor/src/hooks/use-product-variations-helper.ts:45]: Not needed anymore as this has been moved to the `generateVariations` action.

**Review by mdperez86 — APPROVED** [review]
Cool, very good job here @louwie17

**Pull request merged**